gips/0081-indexing-payments.md

```diff
@@ -0,0 +1,165 @@
+
+At the same time, agreements are cancellable by either party at any time, with some limitations put in place to allow indexers to collect for all work done up to that point.
+
+Putting all of this together the solution consists of the following:
+
+- Indexers disclose publicly their price per unit of work.
+- Gateways send indexing agreements to indexers they want to work with.
+- Indexers accept (on-chain) agreements that meet their criteria and get to work indexing.
+- On a regular basis:
+  - Indexers collect for their work, by posting on-chain the original agreement, a Proof of Indexing (PoI) and the amount of work done for that period.
+  - Gateways monitor the QoS per open agreement and assess if they need to cancel (on-chain) and/or replace any of them.
+
+This solution aims to:
+
+- compensate indexers for the exact amount of work they do.
+- allow indexers to gracefully exit the agreement if it can't be fully fulfilled.
+- cap the amount payers will spend on indexing. Though they might spend but get no service back if the cap is too low.
+- incentivize sufficient QoS by awarding agreements to indexers that provide it.
+- provide sufficient transparency and flexibility for gateways, so that they can abstract the complexity and provide simpler pricing to subgraph developers
+
+### Indexing Indexer Selection Algorithm (IISA)
+
+As mentioned above, gateways get to choose who they work with. In Horizon, gateways can be run be any protocol participant and they get to decide how to implement the indexer selection. Nevertheless we propose that a good IISA needs to considers QoS, available stake (with delegation), and other parameters to find the optimal balance between QoS, load-balancing, economic security and price.
+
+### Arbitration and disputes
+
+Since collections are tied to "amount of work done" and this is reported by the indexers themselves, there needs to be a mechanism to punish malicious reporting. We propose here to use slashing. For this to be fair we need the "amount of work done" to be verifiable by an arbitrator in the case of a dispute.
+
+### Amount of work done
+
+As an initial iteration of the "amount of work done" proxy we propose to use indexed blocks and subgraph entities. In particular the amount to be collected by indexers shall be equal to:
+
+- A price per amount of blocks indexed (with a different price for each chain). These are charged incrementally, i.e. every time a PoI is posted, the indexer collects a payment for the new blocks indexed since the last PoI.
+- A price per amount of entities stored by the subgraph (same price for each chain). These are charged on a recurring basis per unit of time, for the total number of entities on the subgraph at the time the PoI is posted times the time since the last PoI.
```

Comment:
```suggestion
- A price per extracted entity of the subgraph (same price for each chain). These are charged on a recurring basis per unit of time, for the total number of entities on the subgraph at the time the PoI is posted times the time since the last PoI.
```